src/server/tiered_storage.cc

```diff
@@ -770,4 +778,114 @@ bool TieredStorage::CanExternalizeEntry(PrimeIterator it) {
+  void SetExternal(std::string_view key, tiering::DiskSegment segment) {
+    if (auto pv = Find(key); pv) {
+      pv->SetIoPending(false);
+      pv->SetExternal(segment.offset, segment.length);  // TODO: Handle memory stats
+    }
+  }
+
+  void ClearIoPending(std::string_view key) {
+    if (auto pv = Find(key); pv)
+      pv->SetIoPending(false);
+  }
+
+  // Find entry by key and store it's up-to-date value in place of external segment
+  void SetInMemory(std::string_view key, std::string_view value) {
+    if (auto pv = Find(key); pv)
+      pv->SetString(value);
+  }
+
+  void ReportStashed(EntryId id, tiering::DiskSegment segment) override {
+    if (holds_alternative<string_view>(id)) {
+      SetExternal(get<string_view>(id), segment);
+    } else {
+      for (const auto& [sub_key, sub_segment] :
+           ts_->bins_->ReportStashed(get<tiering::SmallBins::BinId>(id), segment))
+        SetExternal(string_view{sub_key}, sub_segment);
+    }
+  }
+
+  void ReportFetched(EntryId id, std::string_view value, tiering::DiskSegment segment) override {
+    DCHECK(holds_alternative<string_view>(id));  // we never issue reads for bins
+    SetInMemory(get<string_view>(id), value);
+
+    // Delete value
+    if (segment.length >= 2_KB) {
+      Delete(segment);
+    } else {
+      if (auto bin_segment = ts_->bins_->Delete(segment); bin_segment)
+        Delete(*bin_segment);
+    }
+  }
+
+ private:
+  PrimeValue* Find(std::string_view key) {
+    auto it = db_slice_->FindMutable(DbContext{}, key);
+    return IsValid(it.it) ? &it.it->second : nullptr;
+  }
+
+  TieredStorageV2* ts_;
+  DbSlice* db_slice_;
+};
+
+TieredStorageV2::TieredStorageV2(DbSlice* db_slice)
+    : op_manager_{make_unique<ShardOpManager>(this, db_slice)},
+      bins_{make_unique<tiering::SmallBins>()} {
+}
+
+TieredStorageV2::~TieredStorageV2() {
+}
+
+std::error_code TieredStorageV2::Open(string_view path) {
+  return op_manager_->Open(path);
+}
+
+void TieredStorageV2::Close() {
+  op_manager_->Close();
+}
+
+util::fb2::Future<std::string> TieredStorageV2::Read(string_view key, const PrimeValue& value) {
+  DCHECK(value.IsExternal());
+  return op_manager_->Read(key, value.GetExternalSlice());
+}
+
+void TieredStorageV2::Stash(string_view key, PrimeValue* value) {
+  string buf;
+  string_view value_sv = value->GetSlice(&buf);
+  value->SetIoPending(true);
+
+  if (value->Size() >= 2_KB) {
+    if (auto ec = op_manager_->Stash(key, value_sv); ec)
+      value->SetIoPending(false);
+  } else if (auto bin = bins_->Stash(key, value_sv); bin) {
+    if (auto ec = op_manager_->Stash(bin->first, bin->second); ec) {
+      for (const string& key : bins_->ReportStashAborted(bin->first))
+        op_manager_->ClearIoPending(key);  // clear IO_PENDING flag
+    }
+  }
+}
+void TieredStorageV2::Delete(string_view key, PrimeValue* value) {
+  if (value->IsExternal()) {
+    tiering::DiskSegment segment = value->GetExternalSlice();
+    if (segment.length >= 2_KB) {
```

Comment:
pls name 2_kb with a meaningful name and use it everywhere

Reply:
I will udpate all code in a separate PR, there are many more parts that use inconsistent constants